### PR TITLE
feat: add console, syntax error detection

### PR DESCRIPTION
**Pull request by parkerziegler**
This PR adds a stderr / stdout console to the UI, adds support for early syntax error detection in the Python source, and wires up `stdout` / `stderr` from our Python trace to the console.

### Console UI

The Console UI is quite minimal, taking an array of `Statement`s and rendering them one atop the other. It applies specific formatting for `stderr` as well.

image

Rendering new output is just a matter of building a new `statements` array (more details below).

### Early Syntax Error Detection

If there's a syntax error in the Python source itself, we catch it early, render an error message with the offending line number from the Pyodide error traceback, and disable the Execute button both visually and in the DOM itself (so the user has _no_ mechanism to begin stepping through the code). Here's what that looks like:

image

### Wiring up `stdout` / `stderr` from our Python trace

This is functioning, though it may not be the behavior we want. Right now, we copy everything from `stdout` / `stderr` on all frames to the console. This means that we get the same `stdout` / `stderr` printed multiple times as we step through the code, because frames that pertain to non-`print` calls still carry `stdout` along with them. For example, here's the output for our initial program. Notice that "1", "1 2", "1 2 4" get printed multiple times, because frames carry along `stdout` / `stderr` even if they themselves are not `print` statements.

image

Ideally, we'd be able to flag in the trace that we're executing a `print` statement so we only echo `stdout` once; I'm not sure we'd want to do the same for `stderr` 🤔 I have some ideas on ways to pull this off, but I think we should chat when you have some time. I stuck with the least-opinionated approach for now until we had a chance to chat!

### CODE REVIEW

File: ui/src/App.tsx [hunk 1]
```diff
@@ -131,9 +154,14 @@ const App: React.FC = () => {
           onExecute={executePython}
           onPlayToEnd={onPlayToEnd}
           executing={current.value == 'idle'}
+          error={pythonError}
+        />
+      </div>
+      <div className={styles.panel}>
+        <Console
+          stdout={current.context.data?.[current.context.index]?.stdout || ''}
```
Comment (SybelBlue) [inline ui/src/App.tsx]: debugger-fsm defines a helper method `currentTrace` that would be helpful here:
``` js
currentTrace(current.context?)?.stdout ?? ''
```

File: ui/src/components/Console.tsx [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import TerminalIcon from '../assets/terminal.svg';
+
+import styles from './Console.module.css';
+
+interface Props {
+  stdout: string;
+}
+
+const Console: React.FC<Props> = ({ stdout }) => {
+  const statements = stdout.trim().split('\n');
```
Comment (SybelBlue) [inline ui/src/components/Console.tsx]: We don't want to trim because empty print statements should still be reflected in the output!

**Review by SybelBlue — CHANGES_REQUESTED** [review]
Howdy! This looks great, thanks for the work! 

There's just one problem that you touched on - the current stdout is the full contents of the log at that step of debugging (which is why you see duplicates of the log being printed).

It would be great if the log only showed the current state! There's a couple other tweaks I suggested, but otherwise thanks again!

File: ui/src/components/Console.tsx (hunk 2 above)
Comment (parkerziegler) [reply ui/src/components/Console.tsx]: Ah good catch!

File: ui/src/App.tsx (hunk 1 above)
Comment (parkerziegler) [reply ui/src/App.tsx]: Fixed in latest commit!

File: ui/src/components/Console.tsx (hunk 2 above)
Comment (parkerziegler) [reply ui/src/components/Console.tsx]: Fixed in latest commit!

**Comment by parkerziegler** [comment]
@SybelBlue I think we're all set on the changes now! `stdout` appears in the UI as it shows up in the state machine. I also added `stderr` support. I conditionally render `stderr` since we don't want red >> showing up unless we actually hit some `stderr` in the course of things. I also got to your requested changes. Let me know if there are any last things that we want to try to address in the scope of this PR.

Kapture 2021-11-22 at 17 43 02
.

**Review by SybelBlue — APPROVED** [review]
Awesome work again, thanks!

**Pull request merged**